lib/Doctrine/ORM/Tools/SchemaTool.php

```diff
@@ -274,6 +275,15 @@ public function getSchemaFromMetadata(array $classes)
 
             if (isset($class->table['uniqueConstraints'])) {
                 foreach ($class->table['uniqueConstraints'] as $indexName => $indexData) {
+                    $uniqIndex = new Index($indexName, $indexData['columns'], true, false, [], isset($indexData['options']) ? $indexData['options'] : []);
+
+                    foreach ($table->getIndexes() as $tableIndexName => $tableIndex) {
+                        if ($tableIndex->isFullfilledBy($uniqIndex)) {
+                            $table->dropIndex($tableIndexName);
+                            break;
```

Comment:
@deeky666 It makes sense, but `gatherColumn()`[<sup>1</sup>][id1] is not the only one place where `unique index` can be added.
It could be also added in method `gatherRelationJoinColumns()`[<sup>2</sup>][id2] (for relation `OneToOne`).

About `foreign key`... It creates `index key` (__not__ `unique index`) implicit ONLY when is needed, so when doesn't exist another fulfilling index[<sup>3</sup>][id3]. Therefore when relation is `OneToOne` unique index is created[<sup>2</sup>][id2] and then on creating foreign key normal index is not added[<sup>3</sup>][id3].

My solution is good, because indexes from `uniqueConstraint` are the most important (explicit) so always replaces other (fulfilled) implicit indexes. (Even we have relation `ManyToOne` for a column and `uniqueConstraint` is defined on the column, `index key` created with `foreign key` will be replaced by `unique index` from `uniqueConstraint` - of course, in that case relation should be `OneToOne` and can be defined `uniqueConstraint` to set name for the `unique index`).

Finally, your proposition will be also good, but IMO more complicated, because calls [<sup>1</sup>][id1] or [<sup>2</sup>][id2] should be replaced by new method which will decide if `unique index` should be added or not.

After all, `break;` can be taken out but, in my loop table can have at most one fulfilled index (implicit) (added due column annotation [<sup>1</sup>][id1] or [<sup>2</sup>][id2]).

---
<sup>1</sup> [Add unique index for column with annotation `unique=true` (method `gatherColumn()`)][id1]
<sup>2</sup> [Add unique index for relation `OneToOne` (method `gatherRelationJoinColumns()`)][id2]
<sup>3</sup> [Add `index key` with `foreign key` only when needed][id3]

[id1]: https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/SchemaTool.php#L462 "Add unique index for column with annotation `unique=true` (method `gatherColumn()`)"
[id2]: https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/SchemaTool.php#L680 "Add unique index for relation `OneToOne` (method `gatherRelationJoinColumns()`)"
[id3]: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/Table.php#L556-L560 "Add `index key` with `foreign key` only when needed"